python/construct_portfolio.py

```diff
@@ -0,0 +1,131 @@
+    YIELD community_index, community
+    RETURN community_index, community;
+
+    """
+
+    if number_of_last_n_trading_days <= 2:
+        raise InvalidNumberOfTradingDaysException(
+            "Number of last trading days must be greater than 2."
+        )
+
+    if correlation_measure not in ["pearson", "spearman"]:
+        raise InvalidCorrelationMeasureException(
+            "Correlation measure can only be either pearson or spearman"
+        )
+
+    stock_tickers = np.sort(np.unique(stocks))
+
+    if len(values) == 0 or len(values) < len(stock_tickers) * 3:
+        raise NotEnoughOfDataException(
+            "There need to be atleast three entries of data for each stock"
+        )
+
+    stock_nodes = utils.get_last_n_days(
+        stocks, len(stock_tickers), number_of_last_n_trading_days
+    )
+    stock_trading_values = utils.get_last_n_days(
+        values, len(stock_tickers), number_of_last_n_trading_days
+    )
+
+    stock_nodes = utils.split_data(stock_nodes, number_of_last_n_trading_days)
+
+    sorted_indices = utils.get_sorted_indices(stock_nodes)
+
+    stock_trading_values = utils.split_data_and_sort(
+        stock_trading_values, number_of_last_n_trading_days, sorted_indices
+    )
+
+    correlations = utils.calculate_correlations(
+        stock_trading_values, correlation_measure
+    )
+
+    graph = utils.create_igraph_from_matrix(correlations)
+
+    communities = graph.community_leiden(
+        weights=graph.es["weight"],
+        resolution_parameter=resolution_parameter,
+        n_iterations=-1,
+    )
+
+    return get_records(
+        communities, stock_tickers, stock_trading_values, n_best_performing
+    )
+
+
+def get_records(
+    communities: List[List[int]],
+    stock_tickers: List[str],
+    stock_trading_values: List[float],
+    n_best_performing: int,
+):
+    """Function for creating returning mgp.Record data
+
+    Args:
+        communities (List[List[int]]): List of communities with belonging members indexes
+        stock_tickers (List[str]): stock tickers
+        stock_trading_values (List[float]): stock trading values
+        n_best_performing (int): number of best performing stocks to pick from each community
+
+    Returns:
+        List[mgp.Record(community_index = int, community = str)]: list of mgp.Record(community_index = int, community = str)
+
+    """
+
+    records = []
+    for i, members in enumerate(communities):
+        if n_best_performing > 0:
+            members = utils.get_n_best_performing(
+                stock_trading_values, members, n_best_performing
+            )
+
+        stocks_in_community = ", ".join(stock_tickers[members])
+        records.append(mgp.Record(community_index=i, community=stocks_in_community))
+
+    return records
+
+
+class InvalidNumberOfTradingDaysException(Exception):
+    pass
+
+
+class InvalidCorrelationMeasureException(Exception):
+    pass
+
+
+class NotEnoughOfDataException(Exception):
+    pass
```

Comment:
Problem with `construct_portfolio` module is that it is not a regular query module, as query modules should be generic algorithms (like `igraphalg` below) on top of which we design such python modules as `construct_portfolio`. So you should remove this part from MAGE's query modules, but you should bump `igraphalg.py` as query module, as this is generic.  So you can have module  like `igraph_max_flow` which has function `get`